### PR TITLE
No root struct terminator

### PR DESCRIPTION
**Pull request by jaekwon**
Also, fixes README.
Does not encode "default values", and in nillable lists considers them nil.
This is a checkpoint before diving into proto3 compatibility.

### CODE REVIEW

File: binary-decode.go [hunk 1]
```diff
@@ -599,6 +599,13 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
 		// Read each field.
 		for _, field := range info.Fields {
 
+			// We're done if isRoot & we've consumed all the bytes.
+			// NOTE: any unseen fields are zero, consistent
+			// with the continue statement above.
```
Comment (liamsi) [inline binary-decode.go]: Do you mean the continue statement below? Or another function above (which?)?

File: binary-decode.go [hunk 2]
```diff
@@ -611,7 +618,8 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
 			var fieldNum, typ = uint32(0), Typ3(0x00)
 			fieldNum, typ, _n, err = decodeFieldNumberAndTyp3(bz)
 			if field.BinFieldNum < fieldNum {
-				// Set nil field value.
+				// Set zero field value.
+				// NOTE: See note on isRoot below in this function.
```
Comment (liamsi) [inline binary-decode.go]: Are you referring to the comment above? Or the one on `if !isRoot {`?

File: binary-decode.go [hunk 3]
```diff
@@ -637,24 +645,27 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
...
-		if slide(&bz, &n, _n) && err != nil {
-			return
-		}
-		if typ != Typ3_StructTerm {
-			err = errors.New(fmt.Sprintf("expected StructTerm typ3 byte, got %v", typ))
-			return
+		// If not root object, read "StructTerm".
+		if !isRoot {
+			// NOTE: In the future, we'll need to break out of a loop
+			// when encoutering an StructTerm typ3 byte.
```
Comment (liamsi) [inline binary-decode.go]: s/encoutering/encountering

File: codec.go [hunk 4]
```diff
@@ -79,12 +79,13 @@ type InterfaceOptions struct {
...
 	// These fields are only set when registered (as implementing an interface).
-	Registered       bool        // Registered with RegisterConcrete().
-	PointerPreferred bool        // Deserialize to pointer type if possible.
-	Name             string      // Registered name.
-	Disamb           DisambBytes // Disambiguation bytes derived from name.
-	Prefix           PrefixBytes // Prefix bytes derived from name.
-	ConcreteOptions              // Registration options.
+	Registered       bool // Registered with RegisterConcrete().
+	PointerPreferred bool // Deserialize to pointer type if possible.
+	// NilPreferred     bool        // Deserialize to nil for empty structs if PointerPreferred.
```
Comment (liamsi) [inline codec.go:79]: remove if not used

File: reflect.go [hunk 5]
```diff
@@ -72,14 +72,22 @@ func derefPointers(rv reflect.Value) (drv reflect.Value, isPtr bool, isNilPtr bo
...
 	if isNilPtr {
 		return rv, true
 	} else {
 		switch rv.Kind() {
+		case reflect.Bool:
+			return rv, rv.Bool() == false
+		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+			return rv, rv.Int() == 0
+		case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+			return rv, rv.Uint() == 0
```
Comment (liamsi) [inline reflect.go:114]: We should keep an eye on this for amino_rs too /cc @zmanian This might have some minor implications on the kms code for instance.

**Review by liamsi — COMMENTED** [review]
LGTM with some nitpicks: one/two comments are slightly confusing  (what do they refer to); typo

File: binary-decode.go (hunk 1 above)
Comment (jaekwon) [reply binary-decode.go]: Yeah I removed this in a later PR.

File: reflect.go (hunk 5 above)
Comment (jaekwon) [reply reflect.go:114]: Yeah.  This should be (in later PRs at least) compatible with what Proto3 does.

**Pull request merged**